tinygrad/features/multi.py

```diff
@@ -1,16 +1,43 @@
 from __future__ import annotations
 from typing import Optional, Union, Any, Tuple, List
 import functools, itertools, operator
-from tinygrad.helpers import all_same, dedup, round_up, prod, DEBUG
+from tinygrad.helpers import all_same, all_int, dedup, round_up, prod, DEBUG, getenv
 from tinygrad.dtype import DType, Scalar
 from tinygrad.ops import BinaryOps, LoadOps, UnaryOps, TernaryOps, ReduceOps
 from tinygrad.lazy import LazyBuffer
 from tinygrad.shape.shapetracker import sint
 
-def all_reduce(op:ReduceOps, lbs):
-  # TODO: replace this with ring reduce
+def ring_allreduce(op: ReduceOps, lbs: List[LazyBuffer]):
+  assert all_int(lbs[0].shape), f"does not support symbolic shape {lbs[0].shape}"
+  assert all_same([lb.shape[0] for lb in lbs]), "allreduce with uneven shards is undefined"
   bop = {ReduceOps.SUM:BinaryOps.ADD, ReduceOps.MAX:BinaryOps.MAX}[op]
-  return [functools.reduce(lambda x,y: x.e(bop, y), [x.copy_to_device(lb.device) for x in lbs]) for lb in lbs]
+
+  n_lbs, dim = len(lbs), prod(lbs[0].shape)
+  # Ring allreduce doesn't provide a benefit with only 2 nodes or where number of elements is less than 256k (empirically)
```

Comment:
skipping 2 nodes makes sense. Can you show how you derived the 256k number.

Also what is expected to hit in CI benchmark? I see an improvement in `test/external/external_benchmark_multitensor_allreduce.py` and cifar training. Does it hit llama or resnet?

Reply:
> Can you show how you derived the 256k number.

By running it with different sizes and seeing what's faster

> Also what is expected to hit in CI benchmark? I see an improvement in test/external/external_benchmark_multitensor_allreduce.py and cifar training.

Should be +-12GB/s in benchmark

> Does it hit llama or resnet?

LLaMA no (small allreduce sizes), resnet idk